kubernetes/frontend.yaml

```diff
@@ -12,35 +12,43 @@ spec:
     metadata:
       labels:
         app: frontend
+      annotations:
```

Comment:
I'm not convinced we should change the Prometheus annotations.

Prometheus can be instructed to scrape pods in (at least) two different ways:
- You can annotate pods, which you are doing by adding the annotations to the deployment's pod spec.
- You can annotate the services so that its endpoints are scraped, which the existing version does.

One of the silly things about doing it the first way is that there is no longer any collective identity on the targets being scraped. For example, Prometheus could end up deciding to call the metrics like this:

```
buildbarn_blobaccess_whatever{job="...",instance="bb-frontend-12345",k8s_node="ip-10-11-12-13.eu-west-2.compute.internal"}
```

The problem with that naming is that it doesn't contain any stable label that refers to just bb-frontend. "Just show me the RPC latency for bb-frontend processes" is something you wouldn't be able to answer easily. If you use service endpoint scraping, you can easily add a `k8s_service` label on which you can aggregate.

Reply:
@EdSchouten first time commenter, long time lurker (thanks for this project, it's really great). Your comment so alarmed me that it spurred me to confirm this on my cluster. Prometheus version `2.12.0` and kubernetes version `1.12.10` (both reasonably old, but not ancient). Assuming this hasn't changed since those releases, I looked at the service discovery page for prometheus and it looks like labels are included in the scrape. Would the existing label of `app: frontend` combined with the unique namespace of `buildbarn` not be specific enough to do something like:

```
buildbarn_blobaccess_whatever{__meta_kubernetes_namespace="buildbarn",app="frontend}
```

PS: I clicked on the link to find out more about [nxui](https://nxui.nl) from your github page. I believe your TLS cert is expired.

Reply:
I am happy to keep the service annotations, I was under the impression that annotations could not exist on services after reading https://www.weave.works/docs/cloud/latest/tasks/monitor/configuration-k8s/#per-pod-prometheus-annotations. Upon reading elsewhere, it appears that the source is mistaken or outdated in it's comment 

> These annotations need to be part of the pod metadata. They will have no effect if set on other objects such as Services or DaemonSets. The DaemonSet manifest below will instruct Prometheus to scrape all of its pods on port 9102.

I will revert the annotation changes.

Reply:
Ah, now I understand what’s going on here. Indeed, by using Prometheus labels based on pod labels, you can get a shared identity for containers in a deployment.

That said, many people with larger deployments (including me) tend to disable this feature, because it tends to blow up the number of time series on clusters dramatically (e.g., when making no-op changes to deployments/pods, bumping the modification time label). This is why I resorted to service endpoint scraping at the time.

Let’s stick to scraping service endpoints, because it guarantees a shared identity without any risk of time series blowup.